heritability/rhemc_pipeline.py

```diff
@@ -0,0 +1,1334 @@
+                                     af = ht_anc[snps_out.locus, snps_out.alleles].AF)
+        this_tab = this_tab.annotate(maf = hl.if_else(this_tab.af > 0.5, 1-this_tab.af, this_tab.af))
+        LD_bin_anc = make_quantiles(this_tab.ld_score, nq=args.num_ld_bins, 
+                                    approx=args.approx_quantiles, ht=this_tab)
+        this_tab = this_tab.annotate(maf_bin = discretize(this_tab.maf, MAFRANGE),
+                                     LD_bin = discretize(this_tab.ld_score, LD_bin_anc))
+        this_tab = this_tab.select('maf_bin', 'LD_bin')
+        exploded_tab = explode_bins(this_tab, ['maf_bin','LD_bin'], args.verbose)
+
+        # back up annotations as a hail table
+        exploded_tab.write(path_annot+'ht/'+output_files_annot_noextn[anc] + '.ht',
+                           overwrite=True)
+
+        # output
+        _write_flat_without_key(exploded_tab, path_annot+output_files_annot[anc], 
+                                delimiter=' ', header=False)
+    logging.info('Annotation files created.')
+
+
+def hl_combine_str(*expressions, sep="_") -> hl.StringExpression:
+    """ This function combines StringExpressions. Notably, if a string is missing, 
+    nothing is included.
+    """
+    strings = hl.array(list(expressions))
+    return hl.delimit(strings.filter(lambda x: hl.is_defined(x) & (hl.len(x) > 0)), sep)
+
+    # expressions_replaced = [hl.if_else(hl.is_missing(this_expr), "", this_expr) 
+    #                             for this_expr in expressions]
+    # final_expr = []
+    # for idx,this_expr in enumerate(expressions_replaced):
+    #     if idx == 0:
+    #         final_expr = this_expr
+    #     else:
+    #         tf_1 = hl.len(final_expr) == 0
+    #         tf_2 = hl.len(this_expr) == 0
+    #         final_expr = (hl.case().when(tf_1 & tf_2, final_expr)
+    #                                .when(tf_1, this_expr)
+    #                                .when(tf_2, final_expr)
+    #                                .default(final_expr + sep + this_expr))
+    #         # final_expr = hl.if_else(tf_1 & tf_2, final_expr,
+    #         #                         hl.if_else(tf_1, this_expr,
+    #         #                                    hl.if_else(tf_2, final_expr, 
+    #         #                                               final_expr + sep + this_expr)))
+    # return final_expr
+
+
+def list_precomputed_pheno_files(ancestries):
+    """ Obtain the set of phenotype files that have been pre-formatted for RHEmc.
+
+    Parameters
+    ----------
+    ancestries: `list`
+
+    Returns
+    -------
+    `dict`
+    """
+    dictout = {}
+    for anc in ancestries:
+        thisdir = path_pheno + anc + '/'
+        if hl.hadoop_is_dir(thisdir):
+            dictout.update({anc: [basename(x['path']) for x in hl.hadoop_ls(thisdir)]})
+        else:
+            dictout.update({anc: []})
+    return dictout
+
+
+def list_saige_pheno_files(ancestries):
```

Comment:
To avoid having to redo references to these functions from elsewhere I've created a helper to do the actual dict generation and then the original two functions now just call that one with a function-specific path.